### PR TITLE
Crab V2 Funding & Price Impact Warnings

### PR DESCRIPTION
**Pull request by daryakaviani**
## Deposit Funding Warning

Show deposit warning when:
Current IV < max(historicalIV * 0.9, historicalIV - 0.08)

Screen Shot 2022-07-26 at 12 47 44 PM
Screen Shot 2022-07-26 at 12 48 10 PM 

## Withdraw Funding Warning

Show withdraw warning when:
Current IV > max(historicalIV * 1.1, historicalIV + 0.08)

Screen Shot 2022-07-26 at 12 48 49 PM
Screen Shot 2022-07-26 at 12 49 04 PM

## Price Impact Warnings

Show price impact warning when: 
abs(currentIV - executionIV) > max(0.1 * current IV, 0.08)

Screen Shot 2022-07-26 at 5 59 46 PM
Screen Shot 2022-07-26 at 6 49 41 PM
Screen Shot 2022-07-26 at 6 00 02 PM
Screen Shot 2022-07-27 at 11 24 53 AM

### CODE REVIEW

**Comment by alexisgauba** [comment]
I'm only getting the high price impact warning at around 15% price impact, is it intended to be that high or do we want to surface the warning earlier? @daryakaviani @aleone 

Screen Shot 2022-07-26 at 7 23 48 PM

**Review by alexisgauba — COMMENTED** [review]
I'm only getting the high price impact warning at around 15% price impact, is it intended to be that high or do we want to surface the warning earlier? @daryakaviani @aleone 

Screen Shot 2022-07-26 at 7 23 48 PM

From a UI / UX pov lgtm!

**Comment by daryakaviani** [comment]
> I'm only getting the high price impact warning at around 15% price impact, is it intended to be that high or do we want to surface the warning earlier? @daryakaviani @aleone
> 
> Screen Shot 2022-07-26 at 7 23 48 PM
> 
> From a UI / UX pov lgtm!

Should we decrease the 8%? @aleone

**Comment by aleone** [comment]
> abs(currentIV - executionIV) > max(0.1 * current IV, 0.08)

The reason we are seeing such high slippage is on ropsten squeeth is trading at mark / index ratio of 2.4, which is 428 vol!

So 10% of 428 vol is 42.8 vol, which corresponds to about 15% slippage.  This level of vol is not really reasonable imo.

On mainnet we would see limits of around 1% slippage warnings right now. (10% * 88v = 8.8v, and about 0.09% per vol point, so 0.09% * 8.8v = ~1%)

**Review by alexisgauba — APPROVED** [review]

File: packages/frontend/src/components/Strategies/Crab/CrabTradeV2.tsx [hunk 1]
```diff
@@ -341,14 +410,62 @@ const CrabTradeV2: React.FC<CrabTradeV2Type> = ({ maxCap, depositedAmount }) =>
...
                 <InfoIcon fontSize="medium" />
               </div>
               <Typography variant="caption" color="textSecondary">
                 Crab aims to earn yield in dollar terms. It does this by stacking ETH when price decreases and selling ETH when price increases.
               </Typography>
             </div>
+            { depositFundingWarning && depositOption === 0 ? (
+              <div className={classes.notice}>
+                <div className={classes.infoIcon}>
+                  <Tooltip title={"The strategy sells squeeth to earn yield. Yield is currently lower than usual. You can still deposit, but you may earn less."}>
```
Comment (aleone) [inline packages/frontend/src/components/Strategies/Crab/CrabTradeV2.tsx:424]: Id maybe rephrase this to be "You can still deposit, but you may be more likely to have negative returns or lose money."

Comment (aleone) [reply packages/frontend/src/components/Strategies/Crab/CrabTradeV2.tsx:424]: @alexisgauba

File: packages/frontend/src/components/Strategies/Crab/CrabTradeV2.tsx [hunk 2]
```diff
@@ -146,6 +168,50 @@ const CrabTradeV2: React.FC<CrabTradeV2Type> = ({ maxCap, depositedAmount }) =>
...
+const { depositPriceImpactWarning, withdrawPriceImpactWarning } = useMemo(() => {
+  let depositPriceImpactWarning: Boolean | false
+  let withdrawPriceImpactWarning: Boolean | false
+
+  const squeethPrice = depositOption === 0 ? ethAmountOutFromDeposit.div(squeethAmountInFromDeposit)
+                                           : ethAmountInFromWithdraw.div(squeethAmountOutFromWithdraw)
+  const scalingFactor = new BigNumber(10000)
+  const fundingPeriod = new BigNumber(17.5).div(365)  
+  const executionVol = new BigNumber(Math.log(scalingFactor.times(squeethPrice).div(normFactor.times(ethIndexPrice)).toNumber())).div(fundingPeriod).sqrt()
+  const showPriceImpactWarning = (executionVol.minus(impliedVol)).abs().gt(BigNumber.max(new BigNumber(impliedVol).div(10), 0.08))
```
Comment (aleone) [inline packages/frontend/src/components/Strategies/Crab/CrabTradeV2.tsx:180]: Probably best to use constants for the 10% and the 8%, so can be changed easily in future.

File: packages/frontend/src/components/Strategies/Crab/CrabTradeV2.tsx [hunk 3]
```diff
@@ -146,6 +168,50 @@ const CrabTradeV2: React.FC<CrabTradeV2Type> = ({ maxCap, depositedAmount }) =>
...
+      ethAmountInFromWithdraw,
+      squeethAmountOutFromWithdraw,
+      depositOption,
+  ])
+
+  const { depositFundingWarning, withdrawFundingWarning } = useMemo(() => {
+    let depositFundingWarning: Boolean | false
+    let withdrawFundingWarning: Boolean | false
+
+    const daysInYear = 365
```
Comment (aleone) [inline packages/frontend/src/components/Strategies/Crab/CrabTradeV2.tsx:199]: prb constant would be better

File: packages/frontend/src/components/Strategies/Crab/CrabTradeV2.tsx [hunk 4]
```diff
@@ -146,6 +168,50 @@ const CrabTradeV2: React.FC<CrabTradeV2Type> = ({ maxCap, depositedAmount }) =>
...
+    const impliedVolDiff = new BigNumber(depositOption === 0 ? -0.1 : 0.1)
```
Comment (aleone) [inline packages/frontend/src/components/Strategies/Crab/CrabTradeV2.tsx:200]: ideally constants as well for here, for easy adjustments

**Review by aleone — APPROVED** [review]
few comments, mostly around constants instead of magic #s

i don't fully understand exactly how atoms work, but assuming they are refreshed, it looks like correctly using the live index, live mark, which gives the live funding and vol, and then using execution vol correctly

File: packages/frontend/src/components/Strategies/Crab/CrabTradeV2.tsx (hunk 1 above)
Comment (alexisgauba) [reply packages/frontend/src/components/Strategies/Crab/CrabTradeV2.tsx:424]: good w that!

Comment (alexisgauba) [reply packages/frontend/src/components/Strategies/Crab/CrabTradeV2.tsx:424]:
```suggestion
                  <Tooltip title={"The strategy sells squeeth to earn yield. Yield is currently lower than usual. You can still deposit, but you may be more likely to have negative returns."}>
```

File: packages/frontend/src/components/Strategies/Crab/CrabTradeV2.tsx [hunk 5]
```diff
@@ -146,6 +168,50 @@ const CrabTradeV2: React.FC<CrabTradeV2Type> = ({ maxCap, depositedAmount }) =>
     setStrategyData()
   }, [])
 
+const { depositPriceImpactWarning, withdrawPriceImpactWarning } = useMemo(() => {
+  let depositPriceImpactWarning: Boolean | false
+  let withdrawPriceImpactWarning: Boolean | false
+
+  const squeethPrice = depositOption === 0 ? ethAmountOutFromDeposit.div(squeethAmountInFromDeposit)
```
[the remark this answers is not in the document]
Comment (daryakaviani) [reply packages/frontend/src/components/Strategies/Crab/CrabTradeV2.tsx:175]: Yep!

**Comment by daryakaviani** [comment]
> few comments, mostly around constants instead of magic #s
> 
> i don't fully understand exactly how atoms work, but assuming they are refreshed, it looks like correctly using the live index, live mark, which gives the live funding and vol, and then using execution vol correctly

Yep, atoms are refreshed! [Here's](https://levelup.gitconnected.com/jotai-atom-based-state-management-for-react-1ce8fd380296) how they work.

File: packages/frontend/src/components/Strategies/Crab/CrabTradeV2.tsx [hunk 6]
```diff
@@ -146,6 +168,49 @@ const CrabTradeV2: React.FC<CrabTradeV2Type> = ({ maxCap, depositedAmount }) =>
     setStrategyData()
   }, [])
 
+const { depositPriceImpactWarning, withdrawPriceImpactWarning } = useMemo(() => {
```
Comment (KMKoushik) [inline packages/frontend/src/components/Strategies/Crab/CrabTradeV2.tsx:171]: Change to useAppMemo

File: packages/frontend/src/components/Strategies/Crab/CrabTradeV2.tsx [hunk 7]
```diff
@@ -146,6 +168,49 @@ const CrabTradeV2: React.FC<CrabTradeV2Type> = ({ maxCap, depositedAmount }) =>
...
+      dailyHistoricalFunding.period,
+      impliedVol,
+      ethAmountOutFromDeposit,
+      squeethAmountInFromDeposit,
+      ethAmountInFromWithdraw,
+      squeethAmountOutFromWithdraw,
+      depositOption,
+  ])
+
+  const { depositFundingWarning, withdrawFundingWarning } = useMemo(() => {
```
Comment (KMKoushik) [inline packages/frontend/src/components/Strategies/Crab/CrabTradeV2.tsx:195]: CHange to useAppMemo

**Review by KMKoushik — CHANGES_REQUESTED** [review]

**Comment by KMKoushik** [comment]
Shows this message for .3%.  To reproduce this, go to home page -> reload -> come to strategy page
Screen Shot 2022-07-28 at 9 19 57 am

**Review by KMKoushik — APPROVED** [review]
